src/components/update-service/update-service.controller.ts

```diff
@@ -0,0 +1,355 @@
+      this.configStep.valid = _.get(this.ctrl, 'forms.orderConfigureForm.$valid') && this.configChanged;
+    }, true);
+
+    this.ctrl.nextTitle = 'Update';
+  };
+
+  public showResults = () => {
+    this.clearValidityWatcher();
+    this.ctrl.configPageShown = false;
+    this.ctrl.nextTitle = "Close";
+    this.ctrl.wizardDone = true;
+    this.updateServiceConfig();
+  };
+
+  public selectPlan = (plan: any) => {
+    this.ctrl.selectedPlan = plan;
+
+    if (_.get(plan, 'metadata.name') === _.get(this.ctrl.serviceInstance, 'spec.clusterServicePlanRef.name')) {
+      this.ctrl.parameterData = angular.copy(this.origParameterData);
+    } else {
+      // Clear any previous parameter data since each plan has its own parameter schema.
+      this.ctrl.parameterData = {};
+    }
+    this.updateParameterSchema(plan);
+
+    this.configChanged = !angular.equals(this.ctrl.parameterData, this.origParameterData) || (this.ctrl.selectedPlan !== this.originalPlan);
+    this.configStep.valid = _.get(this.ctrl, 'forms.orderConfigureForm.$valid') && this.configChanged;
+  };
+
+  public updateServiceConfig = () => {
+    this.ctrl.orderComplete = false;
+    this.ctrl.error = null;
+
+    var parameterData: any = this.getParameters(this.ctrl.parameterData);
+
+    // Update the instance parameters
+    var originalParameters: any = _.get(this.ctrl.serviceInstance, 'spec.parameters');
+    var updateParameters: any = _.pick(parameterData, _.keys(originalParameters));
+    var updateInstance: any = angular.copy(this.ctrl.serviceInstance);
+
+    if (_.get(updateInstance, 'spec.externalClusterServicePlanName') !== _.get(this.ctrl.selectedPlan, 'spec.externalName')) {
+      _.set(updateInstance, 'spec.clusterServicePlanRef', undefined);
+      _.set(updateInstance, 'spec.externalClusterServicePlanName', _.get(this.ctrl.selectedPlan, 'spec.externalName'));
+    }
+
+    if (!angular.equals(updateParameters, originalParameters)) {
+      _.set(updateInstance, 'spec.parameters', updateParameters);
+    }
+
+    // Update the secret parameters
+    var updatedSecretParameters: any = {};
+
+    _.each(this.secrets, (secret: any) => {
+      var originalSecretParameters: any = JSON.parse(this.SecretsService.decodeSecretData(secret.data).parameters);
+      var updateSecretParameters: any = _.pick(parameterData, _.keys(originalSecretParameters));
+
+      if (!angular.equals(updateSecretParameters, originalSecretParameters)) {
+        // Add the parameters to the new secret parameters
+        angular.extend(updatedSecretParameters, updateSecretParameters);
+
+        // Remove the reference to this secret
+        updateInstance.spec.parametersFrom = _.reject(updateInstance.spec.parametersFrom, function (parameters: any) {
+          return _.get(parameters, 'secretKeyRef.name') === secret.metadata.name;
+        });
+      }
+    });
+
+    // Add a new secret if any secret parameters were updated
```

Comment:
What happens if a new parameter is added wasn't already set in a secret or in the instance?

Reply:
Also are we handling the case where we're clearing an existing value properly?

Reply:
Clearing works properly, if it was in the parameters it will be removed. if it was in a secret, the secret gets removed from its list and a new secret is added.

Still working on checking for new parameters

Reply:
Added code to add any new parameters to a new secret.